internal/csi-addons/networkfence/fencing.go

```diff
@@ -132,6 +145,268 @@ func (nf *NetworkFence) AddNetworkFence(ctx context.Context) error {
+	ipAddress := net.ParseIP(ip)
+	if ipAddress == nil {
+		log.ErrorLog(ctx, "error parsing IP address %s\n", ip)
+
+		return false
+	}
+
+	// Check if the IP address is within the CIDR block
+	return ipCidr.Contains(ipAddress)
+}
+
+func (ac *activeClient) fetchIP() (string, error) {
+	// example: "inst": "client.4305 172.21.9.34:0/422650892",
+	// then returning value will be 172.21.9.34
+	clientInfo := ac.Inst
+	parts := strings.Fields(clientInfo)
+	if len(parts) >= 2 {
+		ip := strings.Split(parts[1], ":")[0]
+
+		return ip, nil
+	}
+
+	return "", errors.New("failed to extract IP address")
+}
+
+func (ac *activeClient) fetchID() (int, error) {
+	// example: "inst": "client.4305 172.21.9.34:0/422650892",
+	// then returning value will be 4305
+	clientInfo := ac.Inst
+	parts := strings.Fields(clientInfo)
+	if len(parts) >= 1 {
+		clientIDStr := strings.TrimPrefix(parts[0], "client.")
+		clientID, err := strconv.Atoi(clientIDStr)
+		if err != nil {
+			return 0, fmt.Errorf("failed to convert client ID to int: %w", err)
+		}
+
+		return clientID, nil
+	}
+
+	return 0, fmt.Errorf("failed to extract client ID, incorrect format: %s", clientInfo)
+}
+
+// AddClientEviction blocks access for all the IPs in the CIDR block
+// using client eviction.
+// blocks the active clients listed in cidr, and the IPs
+// for whom there is no active client present too.
+func (nf *NetworkFence) AddClientEviction(ctx context.Context) error {
+	evictedIPs := make(map[string]bool)
+	// fetch active clients
+	activeClients, err := nf.listActiveClients(ctx)
+	if err != nil {
+		return err
+	}
+	// iterate through CIDR blocks and check if any active client matches
+	for _, cidr := range nf.Cidr {
+		for _, client := range activeClients {
+			clientIP, err := client.fetchIP()
+			if err != nil {
+				return fmt.Errorf("error fetching client IP: %w", err)
+			}
+			// check if the clientIP is in the CIDR block
+			if isIPInCIDR(ctx, clientIP, cidr) {
+				clientID, err := client.fetchID()
+				if err != nil {
+					return fmt.Errorf("error fetching client ID: %w", err)
+				}
+				// evict the client
+				err = nf.evictCephFSClient(ctx, clientID)
+				if err != nil {
+					return fmt.Errorf("error evicting client %d: %w", clientID, err)
+				}
+				log.DebugLog(ctx, "client %d has been evicted\n", clientID)
+				// add the CIDR to the list of blocklisted IPs
+				evictedIPs[clientIP] = true
+			}
+		}
+	}
+
+	// blocklist the IPs in CIDR without any active clients
+	for _, cidr := range nf.Cidr {
+		// check if the CIDR is evicted
+		// fetch the list of IPs from a CIDR block
+		hosts, err := getIPRange(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to convert CIDR block %s to corresponding IP range: %w", cidr, err)
+		}
+
+		// add ceph blocklist for each IP in the range mentioned by the CIDR
+		for _, host := range hosts {
+			if evictedIPs[host] {
+				continue
+			}
+			err = nf.addCephBlocklist(ctx, host, false)
+			if err != nil {
+				return err
+			}
```

Comment:
i think this can lead to a problem for below case

* Create RBD NetworkFence CR for one IP 10.10.10.10
* Created CephFS NetworkFence CR for IP range but above IP comes in the Range
* Delete the CephFS Network Fence CR to unblocklist the IP

Now the expectation is the IP that is used in RBD NetworkFence is also blocklisted but that's not the case anyone isn't it?

Reply:
Yes, that looks possible.

I do not know if it is possible to see if a blocklisted client is CephFS or RBD. If that is returned (or resolvable otherwise) from listing the blocklisted client output, a check for the type can be added. Otherwise, we'll need to com up with some other way of tracking/identifying.

Reply:
The blocklisted output doesn't have anything to distinguish between two
ex- $ ceph osd blocklist ls
listed 1 entries
127.0.0.1:0/3710147553 2018-03-19 11:32:24.716146

So, one way could be to add a field in NetworkFence struct to track the type of network fence (RBD or CephFS). When creating a NetworkFence object, we can set this field based on the network fence type, and then while evicting we can check if this is cephfs then evict. 
WDYT @Madhu-1 , @nixpanic